### ISSUE TITLE
cant deserialize Map type with key =  0 in NodeJS

### ISSUE THREAD

**Issue by jet6666** [protocolbuffers/protobuf#4500]
`syntax = "proto3" ;
  message MapTest {
		map<int32 ,MapTestItem> test=1 ; 
  }
  
  message MapTestItem {
  	int32 hp=1;
  	int32 hp2=2;
  }`

**Server-side：PHP (protobuf 3.5.0~3.5.1)**
`$p = new MapTest();
$index = 0;　// Index=0 

$item = new MapTestItem();
$item->setHp(99);
 
$p->setToChanged( array($index =>$item) );
$packed=$p->serializeToString();
  
echo "byteLen=",strlen($packed),PHP_EOL;
for($i=0;$i< strlen($packed);$i++) {
  echo bin2hex($packed{$i}),PHP_EOL;
}`

the result goes to :
byteLen=6
0a
04
12
02
08
63

---------------------
But cant deserializeBinary in JS:

`var buffBytes=Buffer.alloc(6);
buffBytes[0] =58;
buffBytes[1] =4;
buffBytes[2] =18;
buffBytes[3] =2;
buffBytes[4] =8;
buffBytes[5] =99;
 function buff2uint(buffBytes) {
	var a = new Uint8Array(buffBytes.length);
	for (var i = 0; i < buffBytes.length; i++) a[i] = buffBytes[i];
		return a ;
}
bytes = buff2uint(buffBytes);
var message2 = proto.MapTest.deserializeBinary(bytes);
`
 
**throw an error:**
`C:\Users\jet\AppData\Roaming\npm\node_modules\google-protobuf\google-protobuf.js:98
goog.string.splitLimit=function(a,b,c){a=a.split(b);for(var d=[];0<c&&a.length;)d.push(a.shift()),c--;a.length&&d.push(a.join(b));return d};goog.string.editDistance=function(a,b){var c=[],d=[];if(a==b)return 0;if(!a.length||!b.length)return Math.max(a.length,b.length);for(var e=0;e<b.length+1;e++)c[e]=e;for(e=0;e<a.length;e++){d[0]=e+1;for(var f=0;f<b.length;f++)d[f+1]=Math.min(d[f]+1,c[f+1]+1,c[f]+Number(a[e]!=b[f]));for(f=0;f<c.length;f++)c[f]=d[f]}return d[b.length]};goog.asserts={};goog.asserts.ENABLE_ASSERTS=goog.DEBUG;goog.asserts.AssertionError=function(a,b){b.unshift(a);goog.debug.Error.call(this,goog.string.subs.apply(null,b));b.shift();this.messagePattern=a};goog.inherits(goog.asserts.AssertionError,goog.debug.Error);goog.asserts.AssertionError.prototype.name="AssertionError";goog.asserts.DEFAULT_ERROR_HANDLER=function(a){throw a;};goog.asserts.errorHandler_=goog.asserts.DEFAULT_ERROR_HANDLER;
                     
AssertionError: Assertion failed
    at new goog.asserts.AssertionError (C:\Users\jet\AppData\Roaming\npm\node_modules\google-protobuf\google-protobuf.js:98:603)
    at Object.goog.asserts.doAssertFailure_ (C:\Users\jet\AppData\Roaming\npm\node_modules\google-protobuf\google-protobuf.js:99:126)
    at Object.goog.asserts.assert (C:\Users\jet\AppData\Roaming\npm\node_modules\google-protobuf\google-protobuf.js:99:385)
    at Function.jspb.Map.deserializeBinary (C:\Users\jet\AppData\Roaming\npm\node_modules\google-protobuf\google-protobuf.js:113:230)`
 

---------------------------------
**But the index >0 will works well** 
`$p->setToChanged( array($index > 0  =>$item) );`

It's a Bug in NodeJS ?

**Comment by TeBoring** [comment]
This should be a problem of js decoder. 0 is a default value for int32. Therefore, in some implementation, it's not encoded. js decoder should be able to handle such case.

**Comment by zEvg** [comment]
Pretty annoying bug. 
Any plans to fix it ? Or maybe some guidance where to fix it for volunteers ?

**Comment by zEvg** [comment]
Actually there is the same problem for value.

Here is an example
**Proto:**
```proto

message TotalDoubleStatisticsResponse {
  map<string, DoubleValues> data = 1;
}

message DoubleValues {
  // metric name -> value
  map<string, double> values = 1;
}
```
**Message example in base64:**
```base64
CjAKBzBmYWI1OGYSJQoVCgpwdmlld0NvdW50EQAAAAAAACZACgwKCnB2aWV3VmFsdWUKCwoHODJiYWE0ORIACgsKBzc1YWE1NmESAA==
```


Here how it's being parsed by `protoc`
```javascript
$ base64 --decode message.base64 | protoc --decode_raw
1 {
  1: "0fab58f"
  2 {
    1 {
      1: "pviewCount"
      2: 0x4026000000000000
    }
    1 {
      1: "pviewValue"
    }
  }
}
1 {
  1: "82baa49"
  2: ""
}
1 {
  1: "75aa56a"
  2: ""
}
```

This is an error I get from NodeJS gRPC client:
```
Error: 13 INTERNAL: Failed to parse server response
    at Object.exports.createStatusError (/project/node_modules/grpc/src/common.js:87:15)
    at Object.onReceiveStatus (/project/node_modules/grpc/src/client_interceptors.js:1188:28)
    at InterceptingListener._callNext (/project/node_modules/grpc/src/client_interceptors.js:564:42)
    at InterceptingListener.onReceiveStatus (/project/node_modules/grpc/src/client_interceptors.js:614:8)
    at callback (/project/node_modules/grpc/src/client_interceptors.js:841:24)
```

**Comment by TeBoring** [comment]
This should have been fixed by #4687.

**Comment by Michael-F-Bryan** [comment]
I believe I'm running into this issue where serializing a `map<uint32, bool>` from binary will fail with an `AssertionError`.

In my case if the server returns `{0: true}` it'll get through to the JavaScript code just fine, but when you return more than one item (i.e. `{0: true, 2: false}`) the assertion will be triggered.

```
uncaught exception: AssertionError: Assertion failed google-protobuf.js:81
    DEFAULT_ERROR_HANDLER google-protobuf.js:81
    doAssertFailure_ google-protobuf.js:82
    assert google-protobuf.js:83
    deserializeBinary google-protobuf.js:411
    deserializeBinaryFromReader server_pb.js:588
    readMessage google-protobuf.js:384
    deserializeBinaryFromReader server_pb.js:587
    readMessage google-protobuf.js:384
    deserializeBinaryFromReader server_pb.js:440
    deserializeBinary server_pb.js:420
    Y index.js:49
    Bb index.js:34
    N index.js:33
    fc index.js:41
    N index.js:40
    K index.js:40
    <anonymous> self-hosted:844
```
